jaxpm/painting.py

```diff
@@ -19,48 +27,106 @@ def cic_paint(mesh, positions, weight=None):
     kernel = 1. - jnp.abs(positions - neighboor_coords)
     kernel = kernel[..., 0] * kernel[..., 1] * kernel[..., 2]
     if weight is not None:
-        kernel = jnp.multiply(jnp.expand_dims(weight, axis=-1), kernel)
+        if jnp.isscalar(weight):
+            kernel = jnp.multiply(jnp.expand_dims(weight, axis=-1), kernel)
+        else:
+            kernel = jnp.multiply(weight.reshape(*positions.shape[:-1]),
+                                  kernel)
 
     neighboor_coords = jnp.mod(
         neighboor_coords.reshape([-1, 8, 3]).astype('int32'),
-        jnp.array(mesh.shape))
+        jnp.array(grid_mesh.shape))
 
     dnums = jax.lax.ScatterDimensionNumbers(update_window_dims=(),
                                             inserted_window_dims=(0, 1, 2),
                                             scatter_dims_to_operand_dims=(0, 1,
                                                                           2))
-    mesh = lax.scatter_add(mesh, neighboor_coords, kernel.reshape([-1, 8]),
-                           dnums)
+    mesh = lax.scatter_add(grid_mesh, neighboor_coords,
+                           kernel.reshape([-1, 8]), dnums)
     return mesh
 
 
-def cic_read(mesh, positions):
+@partial(jax.jit, static_argnums=(3, 4))
+def cic_paint(grid_mesh, positions, weight=None, halo_size=0, sharding=None):
+
+    positions = positions.reshape((*grid_mesh.shape, 3))
+
+    halo_size, halo_extents = get_halo_size(halo_size, sharding)
+    grid_mesh = slice_pad(grid_mesh, halo_size, sharding)
+
+    gpu_mesh = sharding.mesh if sharding is not None else None
+    spec = sharding.spec if sharding is not None else P()
```

Comment:
here I would rather test if `sharding` is an instance of a `NamedSharding` rather than `None`